dnsaas/autocomplete.py

```diff
@@ -0,0 +1,146 @@
+    """
+    Subclasses of this mixin simply need to define a `model` attribute and this
+    mixin will create a queryset using the default manager for that model.
+
+    The `model` attribute must either be a subclass of `django.db.models.Model`
+    or a callable that returns one.
+
+    To specify the custom user model, set `self.model` to
+    `staticmethod(get_user_model)` where `get_user_model` is from
+    `django.contrib.auth`.
+    """
+    def get_queryset(self):
+        if inspect.isclass(self.model) and issubclass(self.model, Model):
+            qs = self.model.objects.all()
+        elif hasattr(self.model, '__call__'):
+            qs = self.model().objects.all()
+        return qs
+
+
+class StaffOnlyAutocompleteMixin(object):
+    """
+    This Autocomplete mixin filters the queryset if the name attribute starts
+    with the query and then calls the next filter_queryset in the MRO (if it
+    exists)
+    """
+    def filter_queryset(self, qs):
+        # The user has to be logged in and a staff member to use this
+        if self.request.user.is_authenticated() and self.request.user.is_staff:
+            super_obj = super(StaffOnlyAutocompleteMixin, self)
+
+            if hasattr(super_obj, 'filter_queryset'):
+                qs = super_obj.filter_queryset(qs)
+        else:
+            # Return no results if the user isn't a staff member
+            qs = qs.none()
+
+        return qs
+
+
+class NameStartswithAutocompleteMixin(object):
+    """
+    This Autocomplete mixin filters the queryset if the name attribute starts
+    with the query and then calls the next filter_queryset in the MRO (if it
+    exists)
+    """
+    def filter_queryset(self, qs):
+        if self.q:
+            qs = qs.filter(name__istartswith=self.q)
+
+        super_obj = super(NameStartswithAutocompleteMixin, self)
+
+        if hasattr(super_obj, 'filter_queryset'):
+            qs = super_obj.filter_queryset(qs)
+        return qs
+
+
+class NameContainsAutocompleteMixin(object):
+    """
+    This Autocomplete mixin filters the queryset if the name attribute contains
+    with the query and then calls the next filter_queryset in the MRO (if it
+    exists)
+    """
+    def filter_queryset(self, qs):
+        if self.q:
+            qs = qs.filter(name__icontains=self.q)
+
+        super_obj = super(NameContainsAutocompleteMixin, self)
+
+        if hasattr(super_obj, 'filter_queryset'):
+            qs = super_obj.filter_queryset(qs)
+        return qs
+
+
+class FilteredAutocompleteMixin(object):
+    """
+    Add mixins to chain filters together.
+
+    SECURITY: MAKE SURE TO FILTER BASED ON THE CURRENT USER!!!
+
+    Note that each filter MUST check its superclass for a `filter_queryset`
+    attribute and then call it for this to work as intended.
+    """
+    def get_queryset(self):
+        super_obj = super(FilteredAutocompleteMixin, self)
+        return self.filter_queryset(super_obj.get_queryset())
```

Comment:
Judging by [this comment](https://github.com/allegro/django-powerdns-dnssec/pull/171#issuecomment-255323590), @xliiv doesn't see the point in having two autocomplete mechanisms, and frankly I agree with him. I'll be removing this in the next rewrite.